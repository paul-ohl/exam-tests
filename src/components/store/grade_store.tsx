import Grade from "components/atoms/Grade.interface";
import { create } from "zustand";
import { devtools, persist } from "zustand/middleware";
// import type { } from '@redux-devtools/extension' // required for devtools typing

interface GradeState {
  grades: Grade[];
  add: (newGrade: Grade) => void;
  remove: (id: number) => void;
  removeAll: () => void;
}

export const useGradeStore = create<GradeState>()(
  devtools(
    persist(
      (set) => ({
        grades: [
          {
            id: 1,
            title: "Maths",
            grade: 12,
            dateAdded: "2021-10-01",
            comment: "TrÃ¨s bon travail",
          },
        ],
        add: (newGrade) =>
          set((state) => ({ grades: [...state.grades, newGrade] })),
        remove: (id) =>
          set((state) => ({
            grades: state.grades.filter((grade) => grade.id !== id),
          })),
        removeAll: () => set({ grades: [] }),
      }),
      {
        name: "bear-storage",
      }
    )
  )
);
