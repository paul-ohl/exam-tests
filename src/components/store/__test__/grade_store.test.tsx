import { test, expect, vi, describe, beforeEach } from "vitest";
import { useGradeStore } from "../grade_store";

describe("Store Zustand grade", () => {
  beforeEach(() => {
    const { removeAll } = useGradeStore.getState();
    removeAll();
  });

  test("Test can add a picture in the picture array", () => {
    const { add } = useGradeStore.getState();

    add({
      id: 1,
      grade: 1,
      title: "test",
      comment: "test",
      dateAdded: "2021-10-10",
    });

    const { grades } = useGradeStore.getState();
    expect(grades).toMatchSnapshot();
  });

  test("Test can add multiple element in the picture array", () => {
    const { add } = useGradeStore.getState();

    add({
      id: 1,
      grade: 1,
      title: "test",
      comment: "test",
      dateAdded: "2021-10-10",
    });
    add({
      id: 2,
      grade: 1,
      title: "test",
      comment: "test",
      dateAdded: "2021-10-10",
    });

    const { grades } = useGradeStore.getState();
    expect(grades).toMatchSnapshot();
  });

  test("Test can delete a picture in the picture array", () => {
    const { add, remove } = useGradeStore.getState();

    add({
      id: 1,
      grade: 1,
      title: "test",
      comment: "test",
      dateAdded: "2021-10-10",
    });
    remove(1);

    const { grades } = useGradeStore.getState();
    expect(grades).toEqual([]);
  });

  test("Test can remove a picture in mutiple picture array", () => {
    const { add, remove } = useGradeStore.getState();

    add({
      id: 1,
      grade: 1,
      title: "test",
      comment: "test",
      dateAdded: "2021-10-10",
    });
    add({
      id: 1,
      grade: 1,
      title: "test",
      comment: "test",
      dateAdded: "2021-10-10",
    });
    remove(1);

    const { grades } = useGradeStore.getState();
    expect(grades).toMatchSnapshot();
  });

  test("Test can remove elt that doesnt exist", () => {
    const { remove } = useGradeStore.getState();
    remove(1);

    const { grades } = useGradeStore.getState();
    expect(grades).toEqual([]);
  });

  test("Test can reset the data inside the store", () => {
    const { add, removeAll } = useGradeStore.getState();

    add({
      id: 1,
      grade: 1,
      title: "test",
      comment: "test",
      dateAdded: "2021-10-10",
    });
    add({
      id: 2,
      grade: 1,
      title: "test",
      comment: "test",
      dateAdded: "2021-10-10",
    });

    removeAll();

    const { grades } = useGradeStore.getState();
    expect(grades).toEqual([]);
  });
});
