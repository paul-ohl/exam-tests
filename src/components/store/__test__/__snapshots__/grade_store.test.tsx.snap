// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Store Zustand grade > Test can add a picture in the picture array 1`] = `
[
  {
    "comment": "test",
    "dateAdded": "2021-10-10",
    "grade": 1,
    "id": 1,
    "title": "test",
  },
]
`;

exports[`Store Zustand grade > Test can add multiple element in the picture array 1`] = `
[
  {
    "comment": "test",
    "dateAdded": "2021-10-10",
    "grade": 1,
    "id": 1,
    "title": "test",
  },
  {
    "comment": "test",
    "dateAdded": "2021-10-10",
    "grade": 1,
    "id": 2,
    "title": "test",
  },
]
`;

exports[`Store Zustand grade > Test can remove a picture in mutiple picture array 1`] = `[]`;
